### PR TITLE
RFC: Implementing `tokio::io::AsyncRead` for `aws_smithy_http::byte_stream::ByteStream`

### PR DESCRIPTION
**Pull request by Velfi**
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by rcoh — COMMENTED** [review]
I think we can impl `From<ByteStream::Error> for `io::Error`? then you can use `StreamReader` right?

File: design/src/rfcs/rfc0017_impl_async_read_for_bytestream.md [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+let data: Box<dyn AsyncRead> = Box::new(some_bytestream);
+// Read a gzipped text file line by line
+for line in BufReader::new(GzipDecoder::new(data))).lines() {
+  println!("{line}");
+}
+```
+
+Users currently have the option of wrapping `ByteStream` in a newtype and implementing the `AsyncRead` trait themselves. This is not a good customer experience for reasons of ergonomics and performance that we'll cover later in this RFC.
+
+## Implementing `AsyncRead` for a newtype
```
Comment (david-perez) [inline design/src/rfcs/rfc0017_impl_async_read_for_bytestream.md:18]: Why can't we implement it directly on `ByteStream`? If we did, a buffered version of `ByteStream` can be created using [`tokio::io::BufReader`](https://docs.rs/tokio/latest/tokio/io/struct.BufReader.html).

**Comment by rcoh** [comment]
the code for this was merged—@zhessler can you close or update?

**Comment by Velfi** [comment]
We merged #1391 so I'm going to close this for now.

**Pull request closed without merging**